text/0000-portability-lint.md

```diff
@@ -0,0 +1,574 @@
+
+    #[cfg(windows)]
+    fn as_raw_handle(&self) -> RawHandle { ... }
+}
+```
+
+If you attempted to call `as_raw_fd`, when compiling on Unix you'd get a warning
+from the portability lint that you're calling an API not available on all
+mainstream platforms. There are basically three ways to react (all of which will
+make the warning go away):
+
+- Decide not to use the API, after discovering that it would reduce portability.
+
+- Decide to use the API, putting the function using it within a `cfg(unix)` as
+  well (which will flag that function as Unix-specific).
+
+- Decide to use the API *in a cross-platform way*, e.g. by providing a Windows
+  version of the same functionality. In that case you can use a new macro,
+  `match_cfg`, to prove to the lint that you are covering all necessary cases.
+
+In code, we'd have:
+
+```rust
+////////////////////////////////////////////////////////////////////////////////
+// The code we might have written initially:
+////////////////////////////////////////////////////////////////////////////////
+
+fn unlabeled() {
+    // Would generate a warning: calling a `unix`-only API while only
+    // assuming a mainstream platform
+    let fd = File::open("foo.txt").unwrap().as_raw_fd();
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Code that opts into platform-specificness:
+////////////////////////////////////////////////////////////////////////////////
+
+#[cfg(unix)]
+fn foo_unix() {
+    // No warning: we're within code that assumes `unix`
+    let fd = File::open("foo.txt").unwrap().as_raw_fd();
+}
+
+#[cfg(windows)]
+fn foo_windows() {
+    // No warning: we're within code that assumes `windows`
+    let handle = File::open("foo.txt").unwrap().as_raw_handle();
+}
+
+#[cfg(linux)]
+fn foo_linux() {
+    // No warning: we're within code that assumes `linux`, which implies `unix`
+    let fd = File::open("foo.txt").unwrap().as_raw_fd();
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Code that provides a cross-platform abstraction
+////////////////////////////////////////////////////////////////////////////////
+
+// No `cfg` label here; it's a cross-platform function.
+fn cross_platform() {
+    // No warning emitted, since handling both `unix` and `windows` means
+    // this code works for all mainstream platforms.
+    match_cfg! {
+        unix => foo_unix(),
+        windows => foo_windows(),
+    }
+}
+```
+
+As with many lints, the portability lint is *best effort*: it is not required to
+provide airtight guarantees about portability. In particular, the proposed
+definition in this RFC ignores trait implementations and dispatch. In practice,
+this hole shouldn't be too significant: it's pretty rare to implement a trait
+conditionally based on platform information, and the goal of this RFC is to
+*guide* code toward mainstream compatibility without imposing a heavy burden.
```

Comment:
Agreed - I feel like we should at least be able to set the lint to error to get _hard_ gurantees, which means not having holes like that in it.